staker/bold_staker.go

```diff
@@ -0,0 +1,378 @@
+	f.Bool(prefix+".api", DefaultBoldConfig.API, "enable api")
+	f.String(prefix+".api-host", DefaultBoldConfig.APIHost, "bold api host")
+	f.Uint16(prefix+".api-port", DefaultBoldConfig.APIPort, "bold api port")
+	f.String(prefix+".api-db-path", DefaultBoldConfig.APIDBPath, "bold api db path")
+	f.StringSlice(prefix+".track-challenge-parent-assertion-hashes", DefaultBoldConfig.TrackChallengeParentAssertionHashes, "only track challenges/edges with these parent assertion hashes")
+}
+
+type BOLDStaker struct {
+	stopwaiter.StopWaiter
+	chalManager        *challengemanager.Manager
+	blockValidator     *BlockValidator
+	rollupAddress      common.Address
+	client             bind.ContractBackend
+	lastWasmModuleRoot common.Hash
+	callOpts           bind.CallOpts
+	validatorConfig    L1ValidatorConfig
+	wallet             ValidatorWalletInterface
+}
+
+func newBOLDStaker(
+	ctx context.Context,
+	validatorConfig L1ValidatorConfig,
+	rollupAddress common.Address,
+	callOpts bind.CallOpts,
+	txOpts *bind.TransactOpts,
+	client arbutil.L1Interface,
+	blockValidator *BlockValidator,
+	statelessBlockValidator *StatelessBlockValidator,
+	config *BoldConfig,
+	dataPoster *dataposter.DataPoster,
+	wallet ValidatorWalletInterface,
+) (*BOLDStaker, error) {
+	manager, err := newBOLDChallengeManager(ctx, rollupAddress, txOpts, client, blockValidator, statelessBlockValidator, config, dataPoster)
+	if err != nil {
+		return nil, err
+	}
+	return &BOLDStaker{
+		chalManager:     manager,
+		blockValidator:  blockValidator,
+		rollupAddress:   rollupAddress,
+		client:          client,
+		callOpts:        callOpts,
+		validatorConfig: validatorConfig,
+		wallet:          wallet,
+	}, nil
+}
+
+func (b *BOLDStaker) Initialize(ctx context.Context) error {
+	if err := b.updateBlockValidatorModuleRoot(ctx); err != nil {
+		return err
+	}
+	walletAddressOrZero := b.wallet.AddressOrZero()
+	if b.blockValidator != nil && b.validatorConfig.StartValidationFromStaked {
+		rollupUserLogic, err := boldrollup.NewRollupUserLogic(b.rollupAddress, b.client)
+		if err != nil {
+			return err
+		}
+		latestStaked, err := rollupUserLogic.LatestStakedAssertion(b.getCallOpts(ctx), walletAddressOrZero)
+		if err != nil {
+			return err
+		}
+		if latestStaked == [32]byte{} {
+			latestConfirmed, err := rollupUserLogic.LatestConfirmed(&bind.CallOpts{Context: ctx})
+			if err != nil {
+				return err
+			}
+			latestStaked = latestConfirmed
+		}
+		assertion, err := readBoldAssertionCreationInfo(
+			ctx,
+			rollupUserLogic,
+			b.client,
+			b.rollupAddress,
+			latestStaked,
+		)
+		if err != nil {
+			return err
+		}
+		afterState := protocol.GoGlobalStateFromSolidity(assertion.AfterState.GlobalState)
+		return b.blockValidator.InitAssumeValid(validator.GoGlobalState(afterState))
+	}
+	return nil
+}
+
+func (b *BOLDStaker) Start(ctxIn context.Context) {
```

Comment:
Notice that the "old staker" also has two loops in "Start" (CallIteratively):
1 that polls the current assertions and send UpdateLatestStaked and UpdateLatestConfirmed notifications.
1 that checks for the wasmModuleRoot in the rollup and updates validator accordingly.

I don't think the new staker has this functionality (and it needs it)

Reply:
I have added the loop for wasmModuleRoot.
Though i am not entirely sure what UpdateLatestStaked and UpdateLatestConfirmed are and how it translates to BOLD, can you please help with a brief around what it is used for ?

Reply:
Idea is to check both what's the latest confirmed assertion and what's the latest assertion this staker is staked on. For any node that didn't stake, finalized and staked will be the same.
API is designed to make it easy for units to to "register" to these events.
Confirmed is mostly used by message pruner. If an assertion was confirmed there won't be a challenge for anything before it and we can permanently delete some data.
Latest staked is mostly used by block-validator. This is mostly considering updates or reboots of the node. The idea there is if you're staked on a node there is no point to validate anything behind it.

Reply:
Done.